File: leetcode/easy/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> pars;
        char t;
        if (s.length() < 2) return false;
        for (char c: s)
        {
            if (isOpening(c))
            {
                pars.push(c);
            } else {
                if (pars.empty())
                    return false;
                t = pars.top();
                pars.pop();
                if (!isMatch(t, c)) {
                    return false;
                }
            }
        }
        if (pars.empty())
            return true;
        return false;
    }
    bool isOpening(char c)
    {
        return c == '(' or c == '[' or c == '{';
    }
    bool isMatch(char t, char c)
    {
        return ((t == '(' && c == ')') or (t == '[' && c == ']') or (t == '{' && c == '}'));
    }
};

File: leetcode/easy/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        op = ["(", "{", "["]
        cl = [")", "}", "]"]

        stack = []

        for i in s:
            if i in op:
                stack.append(i)
            else:
                if len(stack) == 0:
                    return False

                t = stack.pop()

                if op.index(t) != cl.index(i):
                    return False

        if len(stack) != 0:
            return False
        return True
